lib/pos_mempool.go

```diff
@@ -241,11 +270,70 @@ func (mp *PosMempool) Start() error {
 			return errors.Wrapf(err, "PosMempool.Start: Problem loading persisted transactions")
 		}
 	}
+	mp.startAugmentedViewRefreshRoutine()
 
 	mp.status = PosMempoolStatusRunning
 	return nil
 }
 
+func (mp *PosMempool) startAugmentedViewRefreshRoutine() {
+	go func() {
+		for {
+			select {
+			case <-time.After(time.Duration(mp.augmentedBlockViewRefreshIntervalMillis) * time.Millisecond):
+				// If we're not within 10 blocks of the PoS cutover, we don't need to update the
+				// augmentedLatestBlockView.
+				if mp.latestBlockHeight+10 < uint64(mp.params.ForkHeights.ProofOfStake2ConsensusCutoverBlockHeight) {
+					continue
+				}
+				// Update the augmentedLatestBlockView with the latest block view.
+				mp.RLock()
```

Comment:
I don't think you need to hold the lock during the copy. I would rewrite this as:
```
mp.RLock()
readOnlyPointer = mp.readOnlyLatestBlockView
mp.RUnblock()

newView, err := readOnlyPointer.CopyUtxoView()
```

This works because we never modify the value of readOnlyLatestView. Instead, we just change the pointer. So you only need to guard the pointer access, not the copy.

I've always used locks for these cases, but a pointer-copy is so chill I thought *maybe* Go made it so that it's guaranteed to be atomic. And I think on most hardware the pointer-copy IS atomic, meaning you don't even *need* a lock around it. But it looks like it's not a guarantee so best to keep the guard you have there.

This was my research:
https://chat.openai.com/share/f8aade6f-7822-484a-9057-f656f7429669
https://www.reddit.com/r/golang/comments/yb1ijr/race_or_not_race/

Reply:
Update: Resolved this myself.